File: flagfile-ts/src/eval.ts
import { createHash } from 'crypto';
import {
    Atom,
    AstNode,
    ComparisonOp,
    ArrayOp,
    LogicOp,
    MatchOp,
    FnCall,
    Segments,
    atomEquals,
    atomCompare,
    atomToString,
    atomDateTime,
    atomString,
} from './ast.js';

// ── Context type ──────────────────────────────────────────────────

export type Context = Record<string, Atom>;

// ── Variable resolution ───────────────────────────────────────────

function getVariableValueFromContext(
    variable: AstNode,
    context: Context,
): Atom | null {
    switch (variable.type) {
        case 'Variable': {
            const name =
                variable.atom.type === 'Variable' ? variable.atom.value : null;
            if (name === null) return null;
            return context[name] ?? null;
        }
        case 'Constant': {
            if (variable.atom.type === 'Variable') {
                return context[variable.atom.value] ?? null;
            }
            return null;
        }
        case 'Function': {
            if (variable.fn === FnCall.Now) {
                const now = new Date();
                const yyyy = now.getFullYear();
                const mm = String(now.getMonth() + 1).padStart(2, '0');
                const dd = String(now.getDate()).padStart(2, '0');
                const hh = String(now.getHours()).padStart(2, '0');
                const min = String(now.getMinutes()).padStart(2, '0');
                const ss = String(now.getSeconds()).padStart(2, '0');
                return atomDateTime(`${yyyy}-${mm}-${dd}T${hh}:${min}:${ss}`);
            }
            const inner = getVariableValueFromContext(variable.arg, context);
            if (inner === null) return null;
            const s = atomToString(inner);
            switch (variable.fn) {
                case FnCall.Upper:
                    return atomString(s.toUpperCase());
                case FnCall.Lower:
                    return atomString(s.toLowerCase());
                default:
                    return null;
            }
        }
        case 'Coalesce': {
            for (const arg of variable.args) {
                if (arg.type === 'Variable' && arg.atom.type === 'Variable') {
                    const val = context[arg.atom.value];
                    if (val !== undefined) return val;
                } else if (arg.type === 'Constant') {
                    return arg.atom;
                }
            }
            return null;
        }
        default:
            return null;
    }
}

// ── Comparison helper ─────────────────────────────────────────────

function evalComparison(contextVal: Atom, op: ComparisonOp, rhs: Atom): boolean {
    switch (op) {
        case ComparisonOp.Eq:
            return atomEquals(contextVal, rhs);
        case ComparisonOp.NotEq:
            return !atomEquals(contextVal, rhs);
        case ComparisonOp.More:
        case ComparisonOp.MoreEq:
        case ComparisonOp.Less:
        case ComparisonOp.LessEq: {
            const cmp = atomCompare(contextVal, rhs);
            if (cmp === null) return false;
            switch (op) {
                case ComparisonOp.More:   return cmp > 0;
                case ComparisonOp.MoreEq: return cmp >= 0;
                case ComparisonOp.Less:   return cmp < 0;
                case ComparisonOp.LessEq: return cmp <= 0;
            }
        }
    }
}

// ── Main evaluator ────────────────────────────────────────────────

export function evaluate(expr: AstNode, context: Context, flagName?: string, segments?: Segments): boolean {
    switch (expr.type) {
        case 'Constant': {
            if (expr.atom.type === 'Boolean') {
                return expr.atom.value;
            }
            if (expr.atom.type === 'Variable') {
                const val = getVariableValueFromContext(expr, context);
                if (val !== null && val.type === 'Boolean') {
                    return val.value;
                }
            }
            return false;
        }

        case 'Compare': {
            const contextVal = getVariableValueFromContext(expr.left, context);
            const rhsAtom =
                expr.right.type === 'Constant' ? expr.right.atom : null;
            if (contextVal === null || rhsAtom === null) return false;
            return evalComparison(contextVal, expr.op, rhsAtom);
        }

        case 'Match': {
            const contextVal = getVariableValueFromContext(expr.left, context);
            if (contextVal === null) return false;
            const haystack = atomToString(contextVal);
            const rhsAtom = expr.right.type === 'Constant' ? expr.right.atom : null;
            if (rhsAtom === null) return false;
            let matched: boolean;
            if (rhsAtom.type === 'Regex') {
                try {
                    matched = new RegExp(rhsAtom.value).test(haystack);
                } catch {
                    matched = false;
                }
            } else {
                const needle = atomToString(rhsAtom);
                switch (expr.op) {
                    case MatchOp.StartsWith:
                    case MatchOp.NotStartsWith:
                        matched = haystack.startsWith(needle);
                        break;
                    case MatchOp.EndsWith:
                    case MatchOp.NotEndsWith:
                        matched = haystack.endsWith(needle);
                        break;
                    default:
                        matched = haystack.includes(needle);
                        break;
                }
            }
            switch (expr.op) {
                case MatchOp.Contains:
                case MatchOp.StartsWith:
                case MatchOp.EndsWith:
                    return matched;
                case MatchOp.NotContains:
                case MatchOp.NotStartsWith:
                case MatchOp.NotEndsWith:
                    return !matched;
            }
        }

        case 'Array': {
            // Case 1: variable in (literal_list)
            if (expr.right.type === 'List') {
                const varValue = getVariableValueFromContext(expr.left, context);
                if (varValue === null) return false;
                const list = expr.right.items;

                switch (expr.op) {
                    case ArrayOp.In: {
                        for (const item of list) {
                            if (atomEquals(varValue, item)) return true;
                        }
                        return false;
                    }
                    case ArrayOp.NotIn: {
                        for (const item of list) {
                            if (atomEquals(varValue, item)) return false;
                        }
                        return true;
                    }
                }
                return false;
            }
            // Case 2: "literal" in variable (variable resolves to List in context)
            else {
                let searchValue: Atom | null;
                if (expr.left.type === 'Constant' && expr.left.atom.type !== 'Variable') {
                    searchValue = expr.left.atom;
                } else {
                    searchValue = getVariableValueFromContext(expr.left, context);
                }
                const listValue = getVariableValueFromContext(expr.right, context);
                if (searchValue === null || listValue === null || listValue.type !== 'List') {
                    return false;
                }
                const found = listValue.items.some(item => atomEquals(searchValue!, item));
                switch (expr.op) {
                    case ArrayOp.In: return found;
                    case ArrayOp.NotIn: return !found;
                }
                return false;
            }
        }

        case 'Logic': {
            const left = evaluate(expr.left, context, flagName, segments);
            const right = evaluate(expr.right, context, flagName, segments);
            switch (expr.op) {
                case LogicOp.And: return left && right;
                case LogicOp.Or:  return left || right;
            }
            return false;
        }

        case 'Scope': {
            const result = evaluate(expr.expr, context, flagName, segments);
            return expr.negate ? !result : result;
        }

        case 'Segment': {
            if (!segments) return false;
            const segExpr = segments.get(expr.name);
            if (!segExpr) return false;
            return evaluate(segExpr, context, flagName, segments);
        }

        case 'Percentage': {
            const inner = getVariableValueFromContext(expr.field, context);
            if (inner === null) return false;
            const bucketKey = atomToString(inner);
            const flag = flagName ?? 'unknown';
            const input = expr.salt
                ? `${flag}.${expr.salt}.${bucketKey}`
                : `${flag}.${bucketKey}`;
            const hex = createHash('sha1').update(input).digest('hex');
            const substr = hex.slice(0, 15);
            const value = parseInt(substr, 16);
            const bucket = value % 100000;
            const threshold = expr.rate * 1000;
            return bucket < threshold;
        }

        default:
            return false;
    }
}

File: flagfile-go/eval.go
package flagfile

import (
	"crypto/sha1"
	"encoding/hex"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// Context maps variable names to atom values.
type Context map[string]Atom

// Segments maps segment names to their boolean expression.
type Segments map[string]*AstNode

// ContextFrom builds a Context from string values (mirrors CLI test context building).
func ContextFrom(pairs map[string]string) Context {
	ctx := make(Context, len(pairs))
	for k, v := range pairs {
		ctx[k] = AtomFrom(v)
	}
	return ctx
}

// getVariableValue mirrors Rust `get_variable_value_from_context`.
func getVariableValue(node *AstNode, ctx Context) (Atom, bool) {
	if node == nil {
		return Atom{}, false
	}
	switch node.Type {
	case NVariable:
		if node.Atom.Kind == KVariable {
			v, ok := ctx[node.Atom.Str]
			return v, ok
		}
	case NConstant:
		if node.Atom.Kind == KVariable {
			v, ok := ctx[node.Atom.Str]
			return v, ok
		}
	case NFunction:
		switch node.Fn {
		case FnNow:
			return atomDateTime(time.Now().UTC()), true
		default:
			val, ok := getVariableValue(node.Child, ctx)
			if !ok {
				return Atom{}, false
			}
			switch node.Fn {
			case FnUpper:
				return atomString(strings.ToUpper(val.String())), true
			case FnLower:
				return atomString(strings.ToLower(val.String())), true
			}
		}
	case NCoalesce:
		for idx := range node.Args {
			arg := &node.Args[idx]
			switch arg.Type {
			case NVariable:
				if arg.Atom.Kind == KVariable {
					if v, ok := ctx[arg.Atom.Str]; ok {
						return v, true
					}
				}
			case NConstant:
				return arg.Atom, true
			}
		}
		return Atom{}, false
	}
	return Atom{}, false
}

// Eval evaluates an expression without segments.
func Eval(expr *AstNode, ctx Context, flagName string) bool {
	return evalImpl(expr, ctx, flagName, nil)
}

// EvalWithSegments evaluates an expression with segment resolution.
func EvalWithSegments(expr *AstNode, ctx Context, flagName string, segments Segments) bool {
	return evalImpl(expr, ctx, flagName, segments)
}

func evalImpl(expr *AstNode, ctx Context, flagName string, segments Segments) bool {
	if expr == nil {
		return false
	}
	switch expr.Type {
	case NConstant:
		if expr.Atom.Kind == KBoolean {
			return expr.Atom.Bool
		}
		if expr.Atom.Kind == KVariable {
			if v, ok := getVariableValue(expr, ctx); ok && v.Kind == KBoolean {
				return v.Bool
			}
		}
		return false

	case NCompare:
		cval, ok := getVariableValue(expr.Left, ctx)
		if !ok {
			return false
		}
		if expr.Right == nil || expr.Right.Type != NConstant {
			return false
		}
		val := expr.Right.Atom
		switch expr.CompOp {
		case OpEq:
			return atomEq(cval, val)
		case OpNotEq:
			return !atomEq(cval, val)
		default:
			ord, cmpOk := atomCmp(cval, val)
			if !cmpOk {
				return false
			}
			switch expr.CompOp {
			case OpMore:
				return ord > 0
			case OpMoreEq:
				return ord >= 0
			case OpLess:
				return ord < 0
			case OpLessEq:
				return ord <= 0
			}
		}
		return false

	case NArray:
		// Case 1: variable in (literal_list)
		if expr.Right != nil && expr.Right.Type == NList {
			search, ok := getVariableValue(expr.Left, ctx)
			if !ok {
				return false
			}
			switch expr.ArrayOp {
			case ArrayIn:
				for _, it := range expr.Right.List {
					if atomEq(search, it) {
						return true
					}
				}
				return false
			case ArrayNotIn:
				for _, it := range expr.Right.List {
					if atomEq(search, it) {
						return false
					}
				}
				return true
			}
			return false
		}
		// Case 2: "literal" in variable (variable resolves to List)
		var search Atom
		var searchOk bool
		if expr.Left != nil && expr.Left.Type == NConstant && expr.Left.Atom.Kind != KVariable {
			search = expr.Left.Atom
			searchOk = true
		} else {
			search, searchOk = getVariableValue(expr.Left, ctx)
		}
		listVal, listOk := getVariableValue(expr.Right, ctx)
		if !searchOk || !listOk || listVal.Kind != KList {
			return false
		}
		found := false
		for _, it := range listVal.List {
			if atomEq(search, it) {
				found = true
				break
			}
		}
		if expr.ArrayOp == ArrayIn {
			return found
		}
		return !found

	case NMatch:
		cval, ok := getVariableValue(expr.Left, ctx)
		if !ok {
			return false
		}
		haystack := cval.String()
		if expr.Right == nil || expr.Right.Type != NConstant {
			return false
		}
		rhs := expr.Right.Atom
		if rhs.Kind == KRegex {
			matched := false
			if re, err := regexp.Compile(rhs.Str); err == nil {
				matched = re.MatchString(haystack)
			}
			switch expr.MatchOp {
			case MatchContains:
				return matched
			case MatchNotContains:
				return !matched
			default:
				return false
			}
		}
		needle := rhs.String()
		switch expr.MatchOp {
		case MatchContains:
			return strings.Contains(haystack, needle)
		case MatchNotContains:
			return !strings.Contains(haystack, needle)
		case MatchStartsWith:
			return strings.HasPrefix(haystack, needle)
		case MatchNotStartsWith:
			return !strings.HasPrefix(haystack, needle)
		case MatchEndsWith:
			return strings.HasSuffix(haystack, needle)
		case MatchNotEndsWith:
			return !strings.HasSuffix(haystack, needle)
		}
		return false

	case NLogic:
		l := evalImpl(expr.Left, ctx, flagName, segments)
		r := evalImpl(expr.Right, ctx, flagName, segments)
		if expr.LogicOp == LogicAnd {
			return l && r
		}
		return l || r

	case NScope:
		res := evalImpl(expr.Child, ctx, flagName, segments)
		if expr.Negate {
			return !res
		}
		return res

	case NSegment:
		if segments != nil {
			if segExpr, ok := segments[expr.SegmentName]; ok {
				return evalImpl(segExpr, ctx, flagName, segments)
			}
		}
		return false

	case NNullCheck:
		_, ok := getVariableValue(expr.Child, ctx)
		if expr.IsNull {
			return !ok
		}
		return ok

	case NPercentage:
		bucketKey, ok := getVariableValue(expr.Field, ctx)
		if !ok {
			return false
		}
		keyStr := bucketKey.String()
		flag := flagName
		if flag == "" {
			flag = "unknown"
		}
		var input string
		if expr.HasSalt {
			input = flag + "." + expr.Salt + "." + keyStr
		} else {
			input = flag + "." + keyStr
		}
		sum := sha1.Sum([]byte(input))
		hexStr := hex.EncodeToString(sum[:])
		substr := hexStr[:15]
		value, _ := strconv.ParseUint(substr, 16, 64)
		bucket := value % 100000
		threshold := uint64(expr.Rate * 1000.0)
		return bucket < threshold
	}
	return false
}
